arch/xtensa/src/esp32/esp32_rmt.c

```diff
@@ -0,0 +1,315 @@
+
+#ifdef CONFIG_ESP32_RMT
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* RMT methods */
+
+static void esp32rmt_reset(struct rmt_dev_s *dev);
+
+IRAM_ATTR static int esp32rmt_interrupt(int irq, void *context, void *arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32rmt_reset
+ *
+ * Description:
+ *   Reset the RMT device.  Called early to initialize the hardware. This
+ *   function is called, before esp32_rmt_setup().
+ *
+ * Input Parameters:
+ *   dev - An instance of the "upper half" RMT driver state structure.
+ *
+ * Returned Value:
+ *  None
+ *
+ ****************************************************************************/
+
+static void esp32rmt_reset(struct rmt_dev_s *dev)
+{
+  irqstate_t flags;
+  int ret;
+  
+  flags = enter_critical_section();
+
+  modifyreg32(DPORT_PERIP_RST_EN_REG, DPORT_RMT_RST, 1);
+  modifyreg32(DPORT_PERIP_RST_EN_REG, DPORT_RMT_RST, 0);
+  putreg32(0xFFFFFFFF, RMT_INT_CLR_REG); /* Clear any spurious IRQ Flag */
+
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: esp32rmt_setup
+ *
+ * Description:
+ *   Configure the RMT. This method is called the first time that the RMT
+ *   device is opened.  This will occur when the port is first opened.
+ *   This setup includes configuring and attaching RMT interrupts.
+ *
+ * Input Parameters:
+ *   dev - An instance of the "upper half" RMT driver state structure.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int esp32rmt_setup(struct rmt_dev_s *dev)
+{
+  irqstate_t flags;
+  int ret = OK;
+
+  flags = enter_critical_section();
+
+  if (dev->cpuint != -ENOMEM)
+    {
+      /* Disable the provided CPU Interrupt to configure it. */
+      up_disable_irq(dev->cpuint);
+    }
+
+  dev->cpu = up_cpu_index();
+  dev->cpuint = esp32_setup_irq(dev->cpu, dev->periph,
+                                2, ESP32_CPUINT_LEVEL);
```

Comment:
What's the motivation for setting a priority level of **2** here?

Reply:
When setting any value that not **2** I get this:


``
[    0.040000] [CPU1] up_assert: Assertion failed at file:chip/esp32_irq.c line: 607 task: nsh_main
[    0.040000] [CPU1] xtensa_dumpstate: CPU1:
``
Then after some seconds, comes the rest of the stack dump.
I'm really not confortable with using 2 because it is the only value that will actually work. Have you ever encountered such thing when dealing with interrupt priorities?

Reply:
See the comment below: https://github.com/apache/incubator-nuttx/pull/6992/files/d7fd5aaececa41f1c14625701c2871b9cca4cbd3#r961940027

That might be the reason it is crashing.

Reply:
Thanks, 
That was indeed the reason of the crash, however with any priority above 2, the IRQ isn't called on time to re-fill the RMT's buffer, and the ws2812 won't work. Perhaps for another application that uses larger pulses, there will be enough time to re-fill the buffer when using 3, 4 or 5.

In a side note, esp32twai_setup is calling:
``
up_enable_irq(priv->cpuint);
``

Reply:
Since the initial motivation for setting the priority level as 2 no longer exists, let me ask again: what's the motivation for setting a priority level higher than **1**?

> In a side note, esp32twai_setup is calling:
> `up_enable_irq(priv->cpuint);`

It is also wrong.
Could you please submit a new PR with the fix?

Reply:
My mistake, 
I was thinking that 1 was the highest priority. I'll change it to 1.
About the twai, I'll submit the PR later today.

Reply:
> My mistake,
> I was thinking that 1 was the highest priority. I'll change it to 1.

No problem!

> About the twai, I'll submit the PR later today.

Thanks!